pkg/executor/aggregate/agg_spill.go

```diff
@@ -15,15 +15,280 @@
 package aggregate
 
 import (
+	"sync"
 	"sync/atomic"
 
+	"github.com/pingcap/tidb/pkg/executor/aggfuncs"
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/disk"
+	"github.com/pingcap/tidb/pkg/util/hack"
 	"github.com/pingcap/tidb/pkg/util/logutil"
 	"github.com/pingcap/tidb/pkg/util/memory"
 	"go.uber.org/zap"
 )
 
-// maxSpillTimes indicates how many times the data can spill at most.
-const maxSpillTimes = 10
+const (
+	// maxSpillTimes indicates how many times the data can spill at most.
+	maxSpillTimes       = 10
+	partialStageFlag    = 1
+	spilledPartitionNum = 256
+	spillTasksDoneFlag  = -1
+
+	noSpill = iota
+	needSpill
+	inSpilling
+
+	notSpillMode = 0
+	spillMode    = 1
+
+	spillLogInfo string = "memory exceeds quota, set aggregate mode to spill-mode"
+)
+
+type parallelHashAggSpillHelper struct {
+	lock struct {
+		mu               *sync.Mutex
+		cond             *sync.Cond
+		nextPartitionIdx int
+		spilledChunksIO  [][]*chunk.DataInDiskByChunks
+		spillStatus      int
+	}
+
+	partialWorkerNum int
+
+	spillTriggered atomic.Bool
+	memTracker     *memory.Tracker
+	diskTracker    *disk.Tracker
+	hasError       atomic.Bool
+
+	// memory consumption when needSpill flag is set.
+	memoryConsumption atomic.Int64
+	memoryQuota       atomic.Int64
```

Comment:
> 1. We can invoke `mu.Lock()` directly instead of `lock.mu.Lock()`
> 2. Move memoryConsumption and memoryQuota into the mu struct
> 3. cond -> waitIfInSpilling

done